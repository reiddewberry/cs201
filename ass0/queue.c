    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "queue.h"
    #include "dll.h"

    typedef struct queue QUEUE;

    struct QUEUE
        {
        SLL *list;
        void *(display)(void *,FILE *);
        void *(free)(void *);
        };

    QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *)){
        QUEUE *items = malloc(sizeof(QUEUE));
        assert(items != 0);
        items->list = newSLL(d,f);
        items->display = d;
        items->free = f;
        return items;
        }

    void enqueue(QUEUE *items,void *value){
        insertSLL(items->list,items->list->size,value);
        }

    void *dequeue(QUEUE *items){
        assert(items->list->size > 0);
        return removeSLL(items->list,0);
        }

    void *peekQUEUE(QUEUE *items){
        assert(items->list->size > 0);
        return getSLL(items->list,0);
        }

    int sizeQUEUE(QUEUE *items){
        return items->list->size;
        }

    void displayQUEUE(QUEUE *items,FILE *FP){
        fprintf(FP,"<")
        int i = 0;
        for(i=0; i < items->list->size; i++){
            fprintf(FP,getSLL(items->list,i));
            if(i < items->list->size-1){
                fprintf(FP,",")
                }
            }
        fprintf(FP,">")
        }

    void displayQUEUEdebug(QUEUE *items,FILE *FP){
        displaySLLdebug(items->list,FP);
        }

    void freeQUEUE(QUEUE *items){
        freeSLL(items->list);
        free(items);
        }
