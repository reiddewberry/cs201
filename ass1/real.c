#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL
    {
    double value;
    };

REAL *
newREAL(double x)
    {
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
    }

double
getREAL(REAL *v)
    {
    return v->value;
    }

double
setREAL(REAL *v,double x)
    {
    double old = v->value;
    v->value = x;
    return old;
    }

void
displayREAL(void *v,FILE *fp)
    {
    fprintf(fp,"%lf",getREAL((REAL *) v));
    }

int
compareREAL(void *real1,void *real2)
    {
    double temp = getREAL((REAL *) real1) - getREAL((REAL *) real2);
    if(temp < 0)
        {
        return -1;
        }
    if(temp > 0)
        {
        return 1;
        }
    else{
        return 0;
        }
    }

int
compareREALdec(void *real1,void *real2)
    {
    double temp = getREAL((REAL *) real2) - getREAL((REAL *) real1);
    if(temp < 0)
        {
        return -1;
        }
    if(temp > 0)
        {
        return 1;
        }
    else{
        return 0;
        }
    }

void
freeREAL(void *v)
    {
    free((REAL *)v);
    }
