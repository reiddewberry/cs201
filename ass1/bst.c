    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "bst.h"
    #include "queue.h"

    typedef struct bstnode{
        void *data;
        struct bstnode *parent;
        struct bstnode *right;
        struct bstnode *left;
        }BSTNODE;

    BSTNODE *newBSTNODE(void *value){
        BSTNODE *bstNode = malloc(sizeof(BSTNODE));
        assert(bstNode != 0);
        bstNode->data = value;
        bstNode->parent = 0;
        bstNode->right = 0;
        bstNode->left = 0;
        return bstNode;
        }

    void *getBSTNODEvalue(BSTNODE *node){
        return node->data;
        }

    void setBSTNODEvalue(BSTNODE *node,void *value){
        node->data = value;
        }

    BSTNODE *getBSTNODEleft(BSTNODE *node){
        if(node->left == 0){
            return 0;
            }
        return node->left;
        }

    void setBSTNODEleft(BSTNODE *node,BSTNODE *replacement){
        node->left = replacement;
        }

    BSTNODE *getBSTNODEright(BSTNODE *node){
        if(node->right == 0){
            return 0;
            }
        return node->right;
        }

    void setBSTNODEright(BSTNODE *node,BSTNODE *replacement){
        node->right = replacement;
        }

    BSTNODE *getBSTNODEparent(BSTNODE *node){
        if(node->parent == 0){
            return 0;
            }
        return node->parent;
        }

    void setBSTNODEparent(BSTNODE *node,BSTNODE *replacement){
        node->parent = replacement;
        }

    void freeBSTNODE(BSTNODE *node,void (*freefunct)(void *value)){
        if(freefunct != 0){
            freefunct(node->data);
            }
        free(node);
        }

//Binary Search Tree section from here on

    void swapFunct(BSTNODE *move,BSTNODE *dest){
        void *valHolder = move->data;
        move->data = dest->data;
        dest->data = valHolder;
        }

    typedef struct bst{
        BSTNODE *root;
        int size;
        void (*display)(void *, FILE *);
        int (*comp)(void *, void *);
        void (*swap)(BSTNODE *move, BSTNODE *dest);
        void (*free)(void *);
        }BST;

    BST *newBST(
        void (*display)(void *value,FILE *FP),        //display
        int (*comp)(void *val1,void *val2),           //comparator. - lt, 0 eq, +  gt
        void (*swap)(BSTNODE *move,BSTNODE *dest),    //swapper
        void (*free)(void *value)){                   //free
        BST *newBST = malloc(sizeof(BST));
        assert(newBST != 0);
        newBST->root = 0;
        newBST->size = 0;
        newBST->display = display;
        newBST->comp = comp;
        if(swap == 0){
            newBST->swap = swapFunct;
            }
        else{newBST->swap = swap;}
        newBST->free = free;
        return newBST;
        };

    BSTNODE *getBSTroot(BST *tree){
        return tree->root;
        }

    void setBSTroot(BST *tree,BSTNODE *replacement){
        tree->root = replacement;
        }

    void setBSTsize(BST *tree,int size){
        tree->size = size;
        }

    BSTNODE *insertBST(BST *tree,void *value){
        BSTNODE *newNode = malloc(sizeof(BSTNODE));
        assert(newNode != 0);
        newNode->left = 0;
        newNode->right = 0;
        newNode->data = value;
        if(tree->root == 0){
            tree->root = newNode;
            tree->size += 1;
            return newNode;
            }
        BSTNODE *compNode = tree->root;
        while(1){
            if(tree->comp(newNode->data,compNode->data) < 0){
                if(compNode->left == 0){
                    compNode->left = newNode;
                    newNode->parent = compNode;
                    tree->size += 1;
                    return newNode;
                    }
                compNode = compNode->left;
                }
            else{
                if(compNode->right == 0){
                    compNode->right = newNode;
                    newNode->parent = compNode;
                    tree->size += 1;
                    return newNode;
                    }
                compNode = compNode->right;
                }
            }
        }

    BSTNODE *findBST(BST *tree,void *value){
        BSTNODE *compNode = tree->root;
        if(tree->size == 0){
            return 0;
            }
        while(1){
            if(tree->comp(value,compNode->data) < 0){
                if(compNode->left == 0){
                    return 0;
                    }
                compNode = compNode->left;
                }
            else{
                if(tree->comp(value,compNode->data) == 0){
                    return compNode;
                    }
                if(compNode->right == 0){
                    return 0;
                    }
                compNode = compNode->right;
                }
            }
        }

    BSTNODE *deleteBST(BST *tree,void *value){
        BSTNODE *delNode = findBST(tree,value);
        delNode = swapToLeafBST(tree,delNode);
        pruneLeafBST(tree,delNode);
        if(delNode != 0){
            tree->size--;
            }
        return delNode;
        }

    static BSTNODE *getSuccessor(BST *tree,BSTNODE *leafNode){
        BSTNODE *initNode = leafNode;
        while(!(leafNode->left == 0 && leafNode->right == 0)){
            if(leafNode->right != 0){
                if(leafNode->right->left != 0){
                    leafNode = leafNode->right->left;
                    while(leafNode->left != 0){
                        leafNode = leafNode->left;
                        }
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                else{
                    leafNode =leafNode->right;
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                }
            if(leafNode->left != 0){
                leafNode = leafNode->left;
                }
            }
        tree->swap(initNode,leafNode);
        return leafNode;
        }

    static BSTNODE *getPredecessor(BST *tree,BSTNODE *leafNode){
        BSTNODE *initNode = leafNode;
        while(!(leafNode->left == 0 && leafNode->right == 0)){
            if(leafNode->left != 0){
                if(leafNode->left->right != 0){
                    leafNode = leafNode->left->right;
                    while(leafNode->right != 0){
                        leafNode = leafNode->right;
                        }
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                else{
                    leafNode =leafNode->left;
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                }
            if(leafNode->right != 0){
                leafNode = leafNode->right;
                }
            }
        tree->swap(initNode,leafNode);
        return leafNode;
        }

    BSTNODE *swapToLeafBST(BST *tree,BSTNODE *node){
        BSTNODE *leafNode = node;
        if(tree->size == 0 || leafNode == 0){
            return leafNode;
            }
        if(leafNode->right != 0){
            leafNode = getSuccessor(tree,leafNode);
            }
        if(leafNode->left != 0){
            leafNode = getPredecessor(tree,leafNode);
            }
        return leafNode;
        }

    void pruneLeafBST(BST *tree,BSTNODE *leaf){
        if(leaf == 0){
            return;
            }
        if(leaf == tree->root){
            tree->root = 0;
            return;
            }
        if(leaf->parent->right == leaf){
            leaf->parent->right = 0;
            leaf->parent = 0;
            }
        else{
            leaf->parent->left = 0;
            leaf->parent = 0;
            }
        }

    int sizeBST(BST *tree){
        return tree->size;
        }

    static int maxDepthBST(BSTNODE *root){
        if(root == 0){
            return 0;
            }
        else{
            int leftDepth = maxDepthBST(root->left);
            int rightDepth = maxDepthBST(root->right);
            if(leftDepth > rightDepth){
                return leftDepth+1;
                }
            else{
                return rightDepth+1;
                }
            }
        }

    static int minDepthBST(BSTNODE *root){
        if(root->left == 0 || root->right == 0){
            return 1;
            }
        if(root->left == 0){
            return minDepthBST(root->right)+1;
            }
        if(root->right == 0){
            return minDepthBST(root->left)+1;
            }
        if(minDepthBST(root->left) <= minDepthBST(root->right)){
            return minDepthBST(root->left)+1;
            }
        else{
            return minDepthBST(root->right)+1;
            }
        }

    void statisticsBST(BST *tree,FILE *fp){
        fprintf(fp,"Nodes: %d\n", tree->size);
        if(tree->root == 0){
            fprintf(fp,"Minimum depth: -1\n");
            fprintf(fp,"Maximum depth: -1\n");
            return;
            }
        //min depth
        int min = minDepthBST(tree->root);
        fprintf(fp,"Minimum dpeth: %d\n",min);
        //max depth
        int max = maxDepthBST(tree->root);
        fprintf(fp,"Maximum depth: %d\n",max);
        }

    static void preDisplayBST(BST *tree,BSTNODE *printNode,FILE *fp){
        fprintf(fp,"[");
        if(tree->size == 0){
            return;
            }
        tree->display(printNode->data,fp);
        if(printNode->left != 0){
            fprintf(fp," ");
            preDisplayBST(tree,printNode->left,fp);
            fprintf(fp,"]");
            }

        if(printNode->right != 0){
            fprintf(fp," ");
            preDisplayBST(tree,printNode->right,fp);
            fprintf(fp,"]");
            }
        }

    void displayBST(BST *tree,FILE *fp){
        preDisplayBST(tree,tree->root,fp);
        fprintf(fp,"]");
        }

    static void levelDisplayBST(BST *tree,FILE *fp){
        QUEUE *printQ = newQUEUE(0,0);
        enqueue(printQ,tree->root);
        BSTNODE *printNode;
        while(1){
            int sizeQ = sizeQUEUE(printQ);
            if(sizeQ == 0){
                break;
                }
            while(sizeQ > 0){
                printNode = peekQUEUE(printQ);
                tree->display(printNode->data,fp);
                dequeue(printQ);
                if(printNode->left != 0){
                    enqueue(printQ,printNode->left);
                    }
                if(printNode->right != 0){
                    enqueue(printQ,printNode->right);
                    }
                sizeQ -= 1;
                if(sizeQ > 0){
                    fprintf(fp," ");
                    }
                }
            fprintf(fp,"\n");
            }
        freeQUEUE(printQ);
        }

    void displayBSTdebug(BST *tree,FILE *fp){
        if(tree->size == 0){return;}
        levelDisplayBST(tree,fp);
        }

    void freeBST(BST *tree){
        BSTNODE *remNode;
        while(tree->root != 0){
            remNode = swapToLeafBST(tree,tree->root);
            pruneLeafBST(tree,remNode);
            if(tree->free != 0){
                tree->free(remNode->data);
                }
            free(remNode);
            }
        free(tree);
        }
