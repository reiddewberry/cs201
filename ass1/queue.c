/*
Chandler Reid Dewberry
Queue Program
*/
    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "queue.h"
    #include "sll.h"

//QUEUE structure
    typedef struct queue
        {
        SLL *list;                      //pointer to underlying Singly-linked List struct
        void (*display)(void *,FILE *); //passed in display function
        void (*free)(void *);           //passed in free function
        }QUEUE;

/*
Public Function to create and initialize a new QUEUE
Parameters:
    *d - Display function to display whatever is stored in the QUEUE
    *f - Frees whatever is stored in the QUEUE
*/
    QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *)){
        //malloc to allocate memory and assert to verify there is enough
        QUEUE *items = malloc(sizeof(QUEUE));
        assert(items != 0);
        //create and initilize the SLL
        items->list = newSLL(d,f);
        items->display = d;
        items->free = f;
        return items;
        }

//Inserts the passed in value into the back of the SLL
    void enqueue(QUEUE *items,void *value){
        insertSLL(items->list,sizeSLL(items->list),value);
        }

//Removes and Returns what is at the front of the SSL
    void *dequeue(QUEUE *items){
        //assert to verify the list isn't empty and the program is working
        assert(sizeSLL(items->list) > 0);
        return removeSLL(items->list,0);
        }

//Doesn't Remove but Returns the pointer to what is at the front of the SLL
    void *peekQUEUE(QUEUE *items){
        assert(sizeSLL(items->list) > 0);
        return getSLL(items->list,0);
        }

//Returns the # of items in the QUEUE
    int sizeQUEUE(QUEUE *items){
        return sizeSLL(items->list);
        }

/*
Prints the Queue in order
2 Parameters:
    items - the QUEUE to be displayed
    fp - file pointer to print to
ex.
<3,5,2,7>
*/
    void displayQUEUE(QUEUE *items,FILE *FP){
        fprintf(FP,"<");
        int i = 0;
        //iterates throught the SLL by the index #
        for(i=0; i < sizeSLL(items->list); i++){
            items->display(getSLL(items->list,i),FP);
            if(i < sizeSLL(items->list)-1){
                fprintf(FP,",");
                }
            }
        fprintf(FP,">");
        }

//Prints the Queue but with more info
//ex. "head->{3,5,2,7},tail->{7}"
    void displayQUEUEdebug(QUEUE *items,FILE *FP){
        displaySLLdebug(items->list,FP);
        }

/*
Frees: 
    QUEUE
    Underlying SLL
    Data - if a Freeing function was passed in
*/
    void freeQUEUE(QUEUE *items){
        freeSLL(items->list);
        free(items);
        }
