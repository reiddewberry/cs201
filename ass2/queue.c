    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "queue.h"
    #include "sll.h"

    typedef struct queue
        {
        SLL *list;
        void (*display)(void *,FILE *);
        void (*free)(void *);
        }QUEUE;

    QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *)){
        QUEUE *items = malloc(sizeof(QUEUE));
        assert(items != 0);
        items->list = newSLL(d,f);
        items->display = d;
        items->free = f;
        return items;
        }

    void enqueue(QUEUE *items,void *value){
        insertSLL(items->list,sizeSLL(items->list),value);
        }

    void *dequeue(QUEUE *items){
        assert(sizeSLL(items->list) > 0);
        return removeSLL(items->list,0);
        }

    void *peekQUEUE(QUEUE *items){
        assert(sizeSLL(items->list) > 0);
        return getSLL(items->list,0);
        }

    int sizeQUEUE(QUEUE *items){
        return sizeSLL(items->list);
        }

    void displayQUEUE(QUEUE *items,FILE *FP){
        fprintf(FP,"<");
        int i = 0;
        for(i=0; i < sizeSLL(items->list); i++){
            items->display(getSLL(items->list,i),FP);
            if(i < sizeSLL(items->list)-1){
                fprintf(FP,",");
                }
            }
        fprintf(FP,">");
        }

    void displayQUEUEdebug(QUEUE *items,FILE *FP){
        displaySLLdebug(items->list,FP);
        }

    void freeQUEUE(QUEUE *items){
        freeSLL(items->list);
        free(items);
        }
