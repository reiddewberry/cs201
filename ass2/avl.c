    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "bst.h"
    #include "queue.h"

    typedef struct avl{
        void *data;
        }AVL;

    AVL *newAVL(
        void (*)(void *,FILE *),           //display
        int (*)(void *,void *),            //comparator
        void (*)(void *));                 //freeing function


    void insertAVL(AVL *,void *);
    int findAVLcount(AVL *,void *);
    void *findAVL(AVL *,void *);
    void *deleteAVL(AVL *,void *);
    int sizeAVL(AVL *);
    int duplicatesAVL(AVL *);
    void statisticsAVL(AVL *,FILE *);
    void displayAVL(AVL *,FILE *);
    void displayAVLdebug(AVL *,FILE *);
    void freeAVL(AVL *);
