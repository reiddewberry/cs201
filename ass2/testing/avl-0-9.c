#include <stdio.h>
#include <stdlib.h>
#include "integer.h"
#include "real.h"
#include "string.h"
#include "queue.h"
#include "bst.h"
#include "avl.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(110);
    //INTEGER test of AVL, nominal size is 10000
    int i;
    AVL *p = newAVL(displayINTEGER,compareINTEGER,freeINTEGER);
    for (i = 0; i < 10000; ++i)
        {
        int j = random() % 10000;
        INTEGER *a = newINTEGER(j);
        insertAVL(p,a);
        if (findAVLcount(p,a) > 1)
            freeINTEGER(a);
        }
    if (sizeAVL(p) < 200)
        {
        printf("AVL:\n");
        displayAVL(p,stdout);
        printf("debug: ");
        displayAVLdebug(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeAVL(p));
    while (sizeAVL(p) > 200)
        {
        for (i = 0; i < 10000; ++i)
            {
            int j = random() % 10000;
            INTEGER *a = newINTEGER(j);
            int count = findAVLcount(p,a);
            if (count == 1)
                freeINTEGER(deleteAVL(p,a));
            else if (count > 1)
                deleteAVL(p,a);
            freeINTEGER(a);
            }
        }
    if (sizeAVL(p) < 200)
        {
        printf("AVL:\n");
        displayAVL(p,stdout);
        printf("debug: ");
        displayAVLdebug(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeAVL(p));
    statisticsAVL(p,stdout);
    freeAVL(p);
    return 0;
    }
