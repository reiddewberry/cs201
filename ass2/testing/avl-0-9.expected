size is 6296
AVL:
0: 4144-(4144-)X
1: 1855(4144-)L 6478(4144-)R
2: 1210(1855)L 3227[2](1855)R 5676+(6478)L 8235[3](6478)R
3: 638(1210)L 1394[2]-(1210)R 2449[4]-(3227[2])L 3539[2](3227[2])R 5117+(5676+)L 5826(5676+)R 7152[3](8235[3])L 9360(8235[3])R
4: 383+(638)L 998+(638)R 1265[2](1394[2]-)L 1713(1394[2]-)R 2084(2449[4]-)L 2800(2449[4]-)R 3439(3539[2])L 3703-(3539[2])R 4594(5117+)L 5274+(5117+)R 5740-(5826)L 6136[3](5826)R 6967(7152[3])L 7474[2]-(7152[3])R 8794(9360)L 9539(9360)R
5: 209-(383+)L =416[2](383+)R 762[3]-(998+)L =1179[2](998+)R =1231(1265[2])L =1360(1265[2])R 1435-(1713)L 1744-(1713)R =1895(2084)L =2211(2084)R 2482-(2800)L 3017(2800)R 3353[2](3439)L 3463+(3439)R =3578[3](3703-)L 3850[2](3703-)R 4391(4594)L 4725(4594)R 5154[4]-(5274+)L =5336(5274+)R =5688(5740-)L 5797+(5740-)R 5832(6136[3])L 6190[2]-(6136[3])R 6830[2]+(6967)L 7056+(6967)R 7354[2](7474[2]-)L 7766[2](7474[2]-)R 8415[2](8794)L 8902[4](8794)R 9514-(9539)L 9834(9539)R
6: =217[3](209-)R =787(762[3]-)R =1552(1435-)R =1755(1744-)R =2495(2482-)R =2878(3017)L =3028[5](3017)R =3308(3353[2])L =3371(3353[2])R =3445(3463+)L =3799(3850[2])L =4133(3850[2])R 4241[2](4391)L 4436[2]-(4391)R 4656-(4725)L 4970+(4725)R =5166(5154[4]-)R =5748(5797+)L =5830(5832)L =6008(5832)R =6261[2](6190[2]-)R 6576[2](6830[2]+)L =6865(6830[2]+)R 6977-(7056+)L =7088[2](7056+)R =7290(7354[2])L =7445(7354[2])R 7727+(7766[2])L 7903(7766[2])R 8335(8415[2])L 8520(8415[2])R 8849-(8902[4])L 8977[2](8902[4])R =9422(9514-)L 9536+(9514-)R 9600[2]+(9834)L 9918-(9834)R
7: =4208(4241[2])L =4290(4241[2])R =4533[2](4436[2]-)R =4658[3](4656-)R =4824(4970+)L =6531(6576[2])L =6611[2](6576[2])R =6994(6977-)R =7574[2](7727+)L =7788[3](7903)L =8020(7903)R =8275(8335)L =8372(8335)R =8517[2](8520)L =8629[3](8520)R =8865[2](8849-)R =8931(8977[2])L =8990(8977[2])R =9516[2](9536+)L =9553(9600[2]+)L =9975(9918-)R
debug: [4144- [1855 [1210 [638 [383+ [209- [217[3]]] [416[2]]] [998+ [762[3]- [787]] [1179[2]]]] [1394[2]- [1265[2] [1231] [1360]] [1713 [1435- [1552]] [1744- [1755]]]]] [3227[2] [2449[4]- [2084 [1895] [2211]] [2800 [2482- [2495]] [3017 [2878] [3028[5]]]]] [3539[2] [3439 [3353[2] [3308] [3371]] [3463+ [3445]]] [3703- [3578[3]] [3850[2] [3799] [4133]]]]]] [6478 [5676+ [5117+ [4594 [4391 [4241[2] [4208] [4290]] [4436[2]- [4533[2]]]] [4725 [4656- [4658[3]]] [4970+ [4824]]]] [5274+ [5154[4]- [5166]] [5336]]] [5826 [5740- [5688] [5797+ [5748]]] [6136[3] [5832 [5830] [6008]] [6190[2]- [6261[2]]]]]] [8235[3] [7152[3] [6967 [6830[2]+ [6576[2] [6531] [6611[2]]] [6865]] [7056+ [6977- [6994]] [7088[2]]]] [7474[2]- [7354[2] [7290] [7445]] [7766[2] [7727+ [7574[2]]] [7903 [7788[3]] [8020]]]]] [9360 [8794 [8415[2] [8335 [8275] [8372]] [8520 [8517[2]] [8629[3]]]] [8902[4] [8849- [8865[2]]] [8977[2] [8931] [8990]]]] [9539 [9514- [9422] [9536+ [9516[2]]]] [9834 [9600[2]+ [9553]] [9918- [9975]]]]]]]]
size is 121
Duplicates: 58
Nodes: 121
Minimum depth: 5
Maximum depth: 7
